File: rooster-backend/models/PerformanceReview.js
const mongoose = require('mongoose');

const performanceReviewSchema = new mongoose.Schema({
  employeeId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Employee',
    required: true
  },
  employeeName: {
    type: String,
    required: true
  },
  reviewerId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  reviewerName: {
    type: String,
    required: true
  },
  companyId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Company',
    required: true
  },
  reviewPeriod: {
    type: String,
    required: true // e.g., "Q4 2024", "Annual 2024"
  },
  startDate: {
    type: Date,
    required: true
  },
  endDate: {
    type: Date,
    required: true
  },
  dueDate: {
    type: Date,
    required: true
  },
  status: {
    type: String,
    enum: ['draft', 'submitted_for_employee_review', 'employee_review_complete', 'completed', 'overdue'],
    default: 'draft'
  },
  scores: {
    communication: { type: Number, min: 1, max: 5 },
    teamwork: { type: Number, min: 1, max: 5 },
    technical: { type: Number, min: 1, max: 5 },
    leadership: { type: Number, min: 1, max: 5 },
    problemSolving: { type: Number, min: 1, max: 5 },
    initiative: { type: Number, min: 1, max: 5 }
  },
  overallRating: {
    type: Number,
    min: 1,
    max: 5
  },
  comments: {
    type: String
  },
  employeeComments: {
    type: String
  },
  goals: [{
    type: String
  }],
  achievements: [{
    type: String
  }],
  areasOfImprovement: [{
    type: String
  }],
  completedAt: {
    type: Date
  },
  createdBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  updatedBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  }
}, {
  timestamps: true
});

// Index for better performance
performanceReviewSchema.index({ companyId: 1, status: 1 });
performanceReviewSchema.index({ employeeId: 1 });
performanceReviewSchema.index({ reviewerId: 1 });

// Virtual for calculating overall rating from scores
performanceReviewSchema.virtual('calculatedRating').get(function() {
  if (!this.scores) return null;
  
  const scoreValues = Object.values(this.scores).filter(score => score !== null && score !== undefined);
  if (scoreValues.length === 0) return null;
  
  const sum = scoreValues.reduce((acc, score) => acc + score, 0);
  return Math.round((sum / scoreValues.length) * 10) / 10; // Round to 1 decimal place
});

// Method to update status based on due date
performanceReviewSchema.methods.updateStatus = function() {
  if (this.status === 'completed') return this.status;
  
  const now = new Date();
  if (now > this.dueDate && this.status !== 'completed') {
    this.status = 'overdue';
  }
  return this.status;
};

// Pre-save middleware to update status
performanceReviewSchema.pre('save', function(next) {
  this.updateStatus();
  next();
});

// Static method to get statistics for a company
performanceReviewSchema.statics.getCompanyStatistics = async function(companyId) {
  try {
    const stats = await this.aggregate([
      { $match: { companyId: new mongoose.Types.ObjectId(companyId) } },
      {
        $group: {
          _id: null,
          total: { $sum: 1 },
          completed: {
            $sum: { $cond: [{ $eq: ['$status', 'completed'] }, 1, 0] }
          },
          inProgress: {
            $sum: { $cond: [{ $eq: ['$status', 'in_progress'] }, 1, 0] }
          },
          overdue: {
            $sum: { $cond: [{ $eq: ['$status', 'overdue'] }, 1, 0] }
          },
          draft: {
            $sum: { $cond: [{ $eq: ['$status', 'draft'] }, 1, 0] }
          },
          averageRating: { $avg: '$overallRating' }
        }
      }
    ]);
    
    return stats.length > 0 ? stats[0] : {
      total: 0,
      completed: 0,
      inProgress: 0,
      overdue: 0,
      draft: 0,
      averageRating: null
    };
  } catch (error) {
    console.error('Error in getCompanyStatistics:', error);
    throw error;
  }
};

module.exports = mongoose.model('PerformanceReview', performanceReviewSchema); 

File: sns_rooster/lib/models/performance_review.dart
import 'package:flutter/material.dart';

class PerformanceReview {
  final String id;
  final String employeeId;
  final String employeeName;
  final String reviewerId;
  final String reviewerName;
  final String reviewPeriod;
  final DateTime startDate;
  final DateTime endDate;
  final String status; // 'draft', 'in_progress', 'completed', 'overdue'
  final Map<String, dynamic> scores; // Category scores
  final String? comments;
  final String? employeeComments;
  final List<String> goals;
  final List<String> achievements;
  final List<String> areasOfImprovement;
  final double? overallRating;
  final DateTime createdAt;
  final DateTime? completedAt;
  final DateTime dueDate;

  PerformanceReview({
    required this.id,
    required this.employeeId,
    required this.employeeName,
    required this.reviewerId,
    required this.reviewerName,
    required this.reviewPeriod,
    required this.startDate,
    required this.endDate,
    required this.status,
    required this.scores,
    this.comments,
    this.employeeComments,
    required this.goals,
    required this.achievements,
    required this.areasOfImprovement,
    this.overallRating,
    required this.createdAt,
    this.completedAt,
    required this.dueDate,
  });

  factory PerformanceReview.fromJson(Map<String, dynamic> json) {
    // Handle populated employeeId (could be object or string)
    String employeeId;
    String employeeName;
    if (json['employeeId'] is Map<String, dynamic>) {
      employeeId = json['employeeId']['_id'] ?? '';
      // Extract employee name from populated object if available
      final firstName = json['employeeId']['firstName'] ?? '';
      final lastName = json['employeeId']['lastName'] ?? '';
      employeeName = '$firstName $lastName'.trim();
    } else {
      employeeId = json['employeeId'] ?? '';
      employeeName = json['employeeName'] ?? '';
    }

    // Handle populated reviewerId (could be object or string)
    String reviewerId;
    String reviewerName;
    if (json['reviewerId'] is Map<String, dynamic>) {
      reviewerId = json['reviewerId']['_id'] ?? '';
      // Extract reviewer name from populated object if available
      final firstName = json['reviewerId']['firstName'] ?? '';
      final lastName = json['reviewerId']['lastName'] ?? '';
      reviewerName = '$firstName $lastName'.trim();
    } else {
      reviewerId = json['reviewerId'] ?? '';
      reviewerName = json['reviewerName'] ?? '';
    }

    final status = json['status'] ?? 'draft';

    return PerformanceReview(
      id: json['_id'] ?? json['id'] ?? '',
      employeeId: employeeId,
      employeeName: employeeName,
      reviewerId: reviewerId,
      reviewerName: reviewerName,
      reviewPeriod: json['reviewPeriod'] ?? '',
      startDate: DateTime.parse(json['startDate']),
      endDate: DateTime.parse(json['endDate']),
      status: status,
      scores: Map<String, dynamic>.from(json['scores'] ?? {}),
      comments: json['comments'],
      employeeComments: json['employeeComments'],
      goals: List<String>.from(json['goals'] ?? []),
      achievements: List<String>.from(json['achievements'] ?? []),
      areasOfImprovement: List<String>.from(json['areasOfImprovement'] ?? []),
      overallRating: json['overallRating']?.toDouble(),
      createdAt: DateTime.parse(json['createdAt']),
      completedAt: json['completedAt'] != null
          ? DateTime.parse(json['completedAt'])
          : null,
      dueDate: DateTime.parse(json['dueDate']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'employeeId': employeeId,
      'employeeName': employeeName,
      'reviewerId': reviewerId,
      'reviewerName': reviewerName,
      'reviewPeriod': reviewPeriod,
      'startDate': startDate.toIso8601String(),
      'endDate': endDate.toIso8601String(),
      'status': status,
      'scores': scores,
      'comments': comments,
      'employeeComments': employeeComments,
      'goals': goals,
      'achievements': achievements,
      'areasOfImprovement': areasOfImprovement,
      'overallRating': overallRating,
      'createdAt': createdAt.toIso8601String(),
      'completedAt': completedAt?.toIso8601String(),
      'dueDate': dueDate.toIso8601String(),
    };
  }

  String get statusDisplayName {
    switch (status) {
      case 'draft':
        return 'Draft';
      case 'in_progress':
        return 'In Progress';
      case 'submitted_for_employee_review':
        return 'Pending Employee Review';
      case 'employee_review_complete':
        return 'Employee Review Complete';
      case 'completed':
        return 'Completed';
      case 'overdue':
        return 'Overdue';
      default:
        return 'Unknown';
    }
  }

  Color get statusColor {
    switch (status) {
      case 'draft':
        return Colors.grey;
      case 'in_progress':
        return Colors.orange;
      case 'submitted_for_employee_review':
        return Colors.orange;
      case 'employee_review_complete':
        return Colors.blue;
      case 'completed':
        return Colors.green;
      case 'overdue':
        return Colors.red;
      default:
        return Colors.grey;
    }
  }
}
